crates/uv-distribution/src/metadata/mod.rs

```diff
@@ -0,0 +1,401 @@
+            tqdm = { url = invalid url to tqdm-4.66.0-py3-none-any.whl" }
+        "#};
+
+        assert_snapshot!(format_err(input).await, @r###"
+        error: TOML parse error at line 8, column 16
+          |
+        8 | tqdm = { url = invalid url to tqdm-4.66.0-py3-none-any.whl" }
+          |                ^
+        invalid string
+        expected `"`, `'`
+
+        "###);
+    }
+
+    #[tokio::test]
+    async fn invalid_url() {
+        let input = indoc! {r#"
+            [project]
+            name = "foo"
+            version = "0.0.0"
+            dependencies = [
+              "tqdm ==4.66.0",
+            ]
+            [tool.uv.sources]
+            tqdm = { url = "§invalid#+#*Ä" }
+        "#};
+
+        assert_snapshot!(format_err(input).await, @r###"
+        error: TOML parse error at line 8, column 8
+          |
+        8 | tqdm = { url = "§invalid#+#*Ä" }
+          |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^
+        data did not match any variant of untagged enum Source
+
+        "###);
+    }
+
+    #[tokio::test]
+    async fn workspace_and_url_spec() {
+        let input = indoc! {r#"
+            [project]
+            name = "foo"
+            version = "0.0.0"
+            dependencies = [
+              "tqdm @ git+https://github.com/tqdm/tqdm",
+            ]
+            [tool.uv.sources]
+            tqdm = { workspace = true }
+        "#};
+
+        assert_snapshot!(format_err(input).await, @r###"
+        error: Failed to parse entry for: `tqdm`
+          Caused by: Can't combine URLs from both `project.dependencies` and `tool.uv.sources`
+        "###);
+    }
+
+    #[tokio::test]
+    async fn missing_workspace_package() {
+        let input = indoc! {r#"
+            [project]
+            name = "foo"
+            version = "0.0.0"
+            dependencies = [
+              "tqdm ==4.66.0",
+            ]
+            [tool.uv.sources]
+            tqdm = { workspace = true }
+        "#};
+
+        assert_snapshot!(format_err(input).await, @r###"
+        error: Failed to parse entry for: `tqdm`
+          Caused by: Package is not included as workspace package in `tool.uv.workspace`
+        "###);
+    }
+
+    #[tokio::test]
+    async fn cant_be_dynamic() {
+        let input = indoc! {r#"
+            [project]
+            name = "foo"
+            version = "0.0.0"
+            dynamic = [
+                "dependencies"
+            ]
+            [tool.uv.sources]
+            tqdm = { workspace = true }
+        "#};
+
+        assert_snapshot!(format_err(input).await, @r###"
+        error: The following field was marked as dynamic: dependencies
+        "###);
+    }
+
+    #[tokio::test]
+    async fn missing_project_section() {
+        let input = indoc! {"
+            [tool.uv.sources]
+            tqdm = { workspace = true }
+        "};
+
+        assert_snapshot!(format_err(input).await, @r###"
+        error: metadata field project not found
+        "###);
+    }
+}
```

Comment:
(Just moved from `lowering.rs` because it accesses a private API on `Metadata`.)